server/controllers/medical/patients/visits.js

```diff
@@ -178,13 +229,175 @@ function admission(req, res, next) {
     data.patient_uuid = db.bid(data.patient_uuid);
   }
 
-  const sql = `
+  const create = data.hospitalized ? createHospitalization : createVisit;
+  create(data)
+    .then(() => {
+      res.status(201).json({ uuid : visitUuid });
+    })
+    .catch(next)
+    .done();
+}
+
+function createVisit(data) {
+  const visitService = getVisitServiceQuery(data);
+  const visit = _.omit(data, ['bed', 'service']);
+  const { service } = data;
+  const sqlInsertVisit = `
     INSERT INTO patient_visit SET ?;
   `;
+  const transaction = db.transaction();
+  visit.last_service_id = service.id;
+  transaction.addQuery(sqlInsertVisit, [visit]);
+  transaction.addQuery(visitService.query, [visitService.parameters]);
+  return transaction.execute();
+}
+
+function createHospitalization(data) {
+  const visitService = getVisitServiceQuery(data);
+  const { bed, service } = data;
+  const visit = _.omit(data, ['bed', 'service']);
+
+  return Q.fcall(() => {
+    return lookupAutoAvailableBed(bed);
+  })
+    .then(b => {
+      const sqlInsertVisit = `
+        INSERT INTO patient_visit SET ?;
+      `;
+      const sqlInsertHospitalization = `
+        INSERT INTO patient_hospitalization SET ?
+      `;
+      const sqlUpdateBed = `
+        UPDATE bed SET is_occupied = 1 WHERE id = ?;
+      `;
+      const paramInsertHospitalization = {
+        uuid : db.bid(uuid()),
+        patient_visit_uuid : visit.uuid,
+        patient_uuid : visit.patient_uuid,
+        room_uuid : db.bid(bed.room_uuid),
+        bed_id : b.id,
+      };
+      const transaction = db.transaction();
+      // insert a new patient visit
+      visit.last_service_id = service.id;
+      transaction.addQuery(sqlInsertVisit, [visit]);
+      // insert a new patient visit service
+      transaction.addQuery(visitService.query, [visitService.parameters]);
+      // insert a new patient hospitalization
+      transaction.addQuery(sqlInsertHospitalization, [paramInsertHospitalization]);
+      // update the bed for the hopitalized patient
+      transaction.addQuery(sqlUpdateBed, [paramInsertHospitalization.bed_id]);
+      return transaction.execute();
+    });
+}
+
+function getVisitServiceQuery(data) {
+  const { service } = data;
+  const visit = _.omit(data, ['bed', 'service']);
+  const sqlInsertVisitService = `
+    INSERT INTO patient_visit_service SET ?;
+  `;
+  const paramInsertService = {
+    uuid : db.bid(uuid()),
+    service_id : service.id,
+    patient_visit_uuid : visit.uuid,
+  };
+  return {
+    query : sqlInsertVisitService,
+    parameters : paramInsertService,
+  };
+}
 
-  db.exec(sql, [data])
+function lookupAutoAvailableBed(bedOptions) {
```

Comment:
Can we name this "lookupNextAvailableBed"?  We don't really need to know if it was automatic or not in the function name.